env0/resource_variable_set.go

```diff
@@ -0,0 +1,451 @@
+		}
+
+		if variable.Schema.Type == "string" {
+			if len(variable.Schema.Enum) > 0 {
+				ivariable["format"] = "dropdown"
+				ivalues := make([]interface{}, 0)
+				for _, value := range variable.Schema.Enum {
+					ivalues = append(ivalues, value)
+				}
+				ivariable["dropdown_values"] = ivalues
+			} else {
+				ivariable["format"] = "text"
+				ivariable["value"] = variable.Value
+			}
+		} else if variable.Schema.Format == "HCL" {
+			ivariable["format"] = "hcl"
+			ivariable["value"] = variable.Value
+		} else if variable.Schema.Format == "JSON" {
+			ivariable["format"] = "json"
+			ivariable["value"] = variable.Value
+		} else {
+			return nil, fmt.Errorf("unhandled variable use-case: %s", variable.Name)
+		}
+	}
+
+	return res, nil
+}
+
+func getVariablesFromSchema(d *schema.ResourceData, organizationId string) ([]client.ConfigurationVariable, error) {
+	res := []client.ConfigurationVariable{}
+
+	ivariables, ok := d.GetOk("variable")
+	if !ok {
+		return res, nil
+	}
+
+	for _, ivariable := range ivariables.([]interface{}) {
+		variable, err := getVariableFromSchema(ivariable.(map[string]interface{}))
+		if err != nil {
+			return nil, err
+		}
+		variable.OrganizationId = organizationId
+		res = append(res, *variable)
+	}
+
+	return res, nil
+}
+
+func resourceVariableSetCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var err error
+
+	apiClient := meta.(client.ApiClientInterface)
+
+	organizationId, err := apiClient.OrganizationId()
+	if err != nil {
+		return diag.Errorf("failed to get organization id")
+	}
+
+	var payload client.CreateConfigurationSetPayload
+	if err := readResourceData(&payload, d); err != nil {
+		return diag.Errorf("schema resource data deserialization failed: %v", err)
+	}
+
+	if payload.Scope != "organization" && payload.ScopeId == "" {
+		return diag.Errorf("scope_id must be configured for the scope '%s'", payload.Scope)
+	}
+
+	if payload.ConfigurationProperties, err = getVariablesFromSchema(d, organizationId); err != nil {
+		return diag.Errorf("failed to get variables from schema: %v", err)
+	}
+
+	configurationSet, err := apiClient.ConfigurationSetCreate(&payload)
+	if err != nil {
+		return diag.Errorf("failed to create a variable set: %v", err)
+	}
+
+	d.SetId(configurationSet.Id)
+
+	return nil
+}
+
+func resourceVariableSetDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	apiClient := meta.(client.ApiClientInterface)
+
+	id := d.Id()
+
+	if err := apiClient.ConfigurationSetDelete(id); err != nil {
+		return diag.Errorf("failed to delete a variable set: %v", err)
+	}
+
+	return nil
+}
+
+type mergedVariables struct {
+	currentVariables []client.ConfigurationVariable
+	newVariables     []client.ConfigurationVariable
+	deletedVariables []client.ConfigurationVariable
+}
+
+func mergeVariables(schema []client.ConfigurationVariable, api []client.ConfigurationVariable) *mergedVariables {
```

Comment:
again here the same logic should also be in the environment resource can we extract it into a common and use the same un both?

Reply:
The current configuration variables in the environment has more features, but that's not the problem...
The implementation is lacking (bugs/limitations) and the schema is confusing (doesn't align with the UI).
But yes, once we add variables_sets to environment, common code will be used.

About environment configuration - we could deprecate current implementation... and replace it but that's a separate (large) effort...